feat(protocol): PCR-2 fixed base fee

cyberphysic4l opened the pull request:
### DEPRECATED
See https://github.com/iotaledger/iota/pull/5270 for base changes associated with this PR, and supporting PCR-2 PRs

cyberphysic4l committed:
introduce feature flag for fixed base fee

cyberphysic4l committed:
target_rewards -> subsidy and computation_reward -> computation_charge

vekkiokonio committed:
Edit feature flag for fixed_based_fee

vekkiokonio committed:
Modified reference_gas_price() based on protocol config

vekkiokonio committed:
Add base_gas_price to AdvanceEpochParams

cyberphysic4l committed:
Merge branch 'develop' into protocol/pcr2

cyberphysic4l committed:
set fixed_base_fee flag to false in snapshot

muXxer committed:
feat(analytics-indexer): add `computation_cost_burned`

muXxer committed:
feat(core-node): add `computation_cost_burned` to `ChangeEpoch`

muXxer committed:
feat(indexer): add `computation_cost_burned` to checkpoints

muXxer committed:
feat(core-node): add asserts and debug prints for `computation_cost_burned`

muXxer committed:
feat(core-node): add `safe_mode_computation_rewards_burned` to system state

muXxer committed:
feat(framework): update framework-snapshot manifest

cyberphysic4l committed:
Merge branch 'develop' into protocol/pcr2

cyberphysic4l committed:
start adding computation_charge_burned to change epoch

cyberphysic4l committed:
Merge remote-tracking branch 'origin/core-node/feat/add-computation-costs-burned' into protocol/pcr2

cyberphysic4l committed:
implement advance_epoch changes for base fee burning

cyberphysic4l committed:
generate gas cost summary correctly with fixed base fee flag

cyberphysic4l committed:
create IotaSystemStateV2

cyberphysic4l committed:
update to SystemEpochInfoEventV2 with fees and correct burned and minted

cyberphysic4l committed:
update more SystemEpochInfoEventV1 to V2

cyberphysic4l committed:
Merge branch 'develop' into protocol/pcr2

cyberphysic4l committed:
create a v1 of adapter

valeriyr commented on crates/iota-framework/packages/iota-system/sources/iota_system_state_inner.move:
```diff
@@ -108,9 +108,60 @@ module iota_system::iota_system_state_inner {
+        storage_fund: StorageFundV1,
+        /// A list of system config parameters.
+        parameters: SystemParametersV1,
+        /// A capability allows to perform privileged IOTA system operations.
+        iota_system_admin_cap: IotaSystemAdminCap,
+        /// The reference gas price for the current epoch.
+        reference_gas_price: u64,
+        /// A map storing the records of validator reporting each other.
+        /// There is an entry in the map for each validator that has been reported
+        /// at least once. The entry VecSet contains all the validators that reported
+        /// them. If a validator has never been reported they don't have an entry in this map.
+        /// This map persists across epoch: a peer continues being in a reported state until the
+        /// reporter doesn't explicitly remove their report.
+        /// Note that in case we want to support validator address change in future,
+        /// the reports should be based on validator ids
+        validator_report_records: VecMap<address, VecSet<address>>,
+
+        /// Whether the system is running in a downgraded safe mode due to a non-recoverable bug.
+        /// This is set whenever we failed to execute advance_epoch, and ended up executing advance_epoch_safe_mode.
+        /// It can be reset once we are able to successfully execute advance_epoch.
+        /// The rest of the fields starting with `safe_mode_` are accmulated during safe mode
+        /// when advance_epoch_safe_mode is executed. They will eventually be processed once we
+        /// are out of safe mode.
+        safe_mode: bool,
+        safe_mode_storage_charges: Balance<IOTA>,
+        safe_mode_computation_charges: Balance<IOTA>,
+        safe_mode_computation_charges_burned: u64,
+        safe_mode_storage_rebates: u64,
+        safe_mode_non_refundable_storage_fee: u64,
+
+        /// Unix timestamp of the current epoch start
+        epoch_start_timestamp_ms: u64,
+        /// Any extra fields that's not defined statically.
+        extra_fields: Bag,
+    }
+
     /// Event containing system-level epoch information, emitted during
     /// the epoch advancement transaction.
-    public struct SystemEpochInfoEventV1 has copy, drop {
+    public struct SystemEpochInfoEventV has copy, drop {
```
Seems like since the framework is updated the first event version can be removed from it because it will not be emitted. Or the name should be reverted.

valeriyr commented on iota-execution/v1/iota-adapter/Cargo.toml:
```diff
@@ -0,0 +1,46 @@
+[package]
```
The `v1` execution layer cut should be created by the following command:
```sh
./scripts/execution_layer.py cut v1
```

I would suggest:
1. Remove the created cut.
2. Revert the `latest` cut changes.
3. Create a new `v1` cut using the script. At this step the `v1` and `latest` cuts are equal, so we freeze the execution layer behavior. Since the `v1` cut is created it should not be modified. The situations when it is possible are described [here](https://github.com/iotaledger/iota/blob/8f82d05d29811e4829effffce13d9224a48fb6c0/iota-execution/README.md#L86-L103)
5. Apply the required changes to the `latest` cut.

valeriyr commented on crates/iota-core/src/authority.rs:
```diff
@@ -1951,6 +1952,7 @@ impl AuthorityState {
         } else {
             transaction.gas().to_vec()
         };
+        let reference_gas_price = epoch_store.reference_gas_price();
```
A variable with the same name is declared earlier in the same function.
It is a bit misleading if we need this declaration or if is it added accidentally.
I would suggest adding a comment if it is required to retake the value here.

cyberphysic4l replied:
That was added accidentally

valeriyr commented on crates/iota-core/src/authority.rs:
```diff
@@ -4511,7 +4513,7 @@ impl AuthorityState {
     ) -> anyhow::Result<(
         IotaSystemState,
-        Option<SystemEpochInfoEventV1>,
+        Option<SystemEpochInfoEventV2>,
```
If we have two system epoch info event versions it means that depending on the protocol version the structure which is returned here can be different.
Probably in this case, we need to consider creating an enum like the following and return it:
```rust
pub enum SystemEpochInfoEvent {
    V1(SystemEpochInfoEventV1),
    V2(SystemEpochInfoEventV2),
}
```

valeriyr commented on crates/iota-framework-snapshot/bytecode_snapshot/1/0x0000000000000000000000000000000000000000000000000000000000000003:
The `1` protocol config snapshots shouldn't be changed, the `2` folder with new snapshots should be created instead.

It will be created automatically by the `./scripts/update_all_snapshots.sh` script once the maximum protocol version is updated.

valeriyr commented on crates/iota-protocol-config/src/lib.rs:
```diff
@@ -585,12 +589,15 @@ pub struct ProtocolConfig {
     /// The number of tokens that the set of validators should receive per
     /// epoch.
-    validator_target_reward: Option<u64>,
+    validator_subsidy: Option<u64>,
```
It is not allowed to rename the parameters, I think. Changes must be additive.

valeriyr commented on crates/iota-protocol-config/src/lib.rs:
```diff
@@ -1583,6 +1596,7 @@ impl ProtocolConfig {
         };
 
         cfg.feature_flags.consensus_transaction_ordering = ConsensusTransactionOrdering::ByGasPrice;
+        cfg.feature_flags.fixed_base_fee = true;
```
The already created protocol configs must stay constant, so if we need to add/change a parameter we need to create a new protocol config version.

The execution layer version also must be changed to `2` in a new protocol config.

valeriyr commented on crates/iota-protocol-config/src/snapshots/iota_protocol_config__test__Mainnet_version_1.snap:
```diff
@@ -7,6 +7,7 @@ feature_flags:
   consensus_transaction_ordering: ByGasPrice
```
This file shouldn't be changed.

valeriyr commented on crates/iota-protocol-config/src/snapshots/iota_protocol_config__test__Testnet_version_1.snap:
```diff
@@ -7,6 +7,7 @@ feature_flags:
   consensus_transaction_ordering: ByGasPrice
```
This file shouldn't be changed.

valeriyr commented on crates/iota-protocol-config/src/snapshots/iota_protocol_config__test__version_1.snap:
```diff
@@ -11,6 +11,7 @@ feature_flags:
   zklogin_max_epoch_upper_bound_delta: 30
```
This file shouldn't be changed.

valeriyr commented on crates/iota-types/src/iota_system_state/iota_system_state_inner_v2.rs:
```diff
@@ -0,0 +1,290 @@
+/// Rust version of the Move
+/// iota_system::validator_cap::UnverifiedValidatorOperationCap type
+#[derive(Debug, Serialize, Deserialize, Clone, Eq, PartialEq)]
+pub struct UnverifiedValidatorOperationCap {
```
This struct can be removed, I think.

valeriyr commented on crates/transaction-fuzzer/src/account_universe/transfer_gen.rs:
```diff
@@ -422,6 +422,11 @@ impl AUTransactionGen for P2PTransferGenRandomGasRandomPriceRandomSponsorship {
         // *sender.current_balances.last().unwrap();
         let rgp = exec.get_reference_gas_price();
         let run_info = RunInfo::new(gas_balance, rgp, self);
+        let reference_gas_price = if PROTOCOL_CONFIG.fixed_base_fee() {
+            PROTOCOL_CONFIG.base_gas_price()
+        } else {
+            exec.get_reference_gas_price()
+        };
```
This code should be extracted in a function and used where it is needed because it is already duplicated in the authority per epoch store.

valeriyr commented on crates/iota-types/src/iota_system_state/mod.rs:
```diff
@@ -190,6 +192,7 @@ pub trait IotaSystemStateTrait {
 #[enum_dispatch(IotaSystemStateTrait)]
 pub enum IotaSystemState {
     V1(IotaSystemStateV1),
+    V2(IotaSystemStateV2),
```
The second system state version also needs to be added in the [advance_epoch_safe_mode](https://github.com/iotaledger/iota/blob/0d6616b33284cad5638687f156aaad30aa1d6138/crates/iota-types/src/iota_system_state/mod.rs#L85) function.

cyberphysic4l committed:
fixes from valerii review

cyberphysic4l committed:
create v1 cut of adapter using script

cyberphysic4l committed:
revert snapshot changes

cyberphysic4l committed:
create v2 framework snapshot

cyberphysic4l committed:
revert v1 snapshot changes

valeriyr committed:
fix: the system framework backward compatibility issue fixed and snapshots updated

valeriyr committed:
fix(iota-protocol-config): fixed resolving protocol version for simulation tests

cyberphysic4l committed:
update mock iota systems

cyberphysic4l committed:
add tests for tips to iota-system

cyberphysic4l committed:
format cargo.toml files

miker83z commented on crates/iota-types/src/gas_model/gas_v2.rs:
```diff
@@ -0,0 +1,522 @@
+        /// remaining gas and storage cost to derive computation cost.
+        fn summary(&self) -> GasCostSummary {
+            // compute computation cost burned and storage rebate, both rebate and non refundable fee
+            let computation_cost_burned = self.computation_cost * self.reference_gas_price / self.gas_price;
```
If this is the only change, then we don't need gas_v2. Maybe an `if fixed_base_fee` in gas_v1 would be sufficient

miker83z replied:
@cyberphysic4l pls, check also this!

cyberphysic4l replied:
I have reverted the addition of gas_v2, and in fact I don't think we even need an `if fixed_base_fee` because the computation_cost_burned field is actually unused in protocol v1, so it doesn't matter what we set for this value. So we avoid an additional if statement here.

cyberphysic4l committed:
rustfmt

cyberphysic4l committed:
clippy lint

cyberphysic4l committed:
Merge branch 'develop' into protocol/pcr2

valeriyr commented on iota-execution/latest/iota-adapter/src/execution_engine.rs:
```diff
@@ -784,18 +785,19 @@ mod checked {
+            computation_charges,
         ];
         let call_arg_arguments = vec![
+            CallArg::Pure(bcs::to_bytes(&params.computation_charge_burned).unwrap()),
```
If it is the only change in the executor we can consider options instead of creating a new execution layer cut, for example:
1. Create a new transaction type `EndOfEpochTransactionKind::ChangeEpochV2`.
2. Or even easier, add a feature flag in the protocol config `advance_epoch_v2`, set it to `true` for the second protocol version, and add the `computation_charge_burned` parameter if it is enabled.

Creating a cut adds a pretty noticeable amount of code to the repository and should be avoided if it is possible, I believe. Because this increases the repository size and build time.

miker83z replied:
I think too we can definitively try avoid the use of a new cut!

cyberphysic4l replied:
Ok, if that is the preferred approach I can try one of the suggested alternatives.

cyberphysic4l committed:
fix  pt_fuzz test

miker83z commented on crates/iota-types/src/gas_model/gas_v1.rs:
```diff
@@ -372,24 +373,20 @@ mod checked {
             } else {
-                self.computation_cost = gas_used;
+                self.computation_cost = computation_cost;
                 Ok(())
```
LGTM!

cyberphysic4l committed:
remove need for gas_v2 and adapter cut v1

cyberphysic4l committed:
Revert "create v1 cut of adapter using script"

This reverts commit 16d8c52b730319e3375fd0764eed96a0244dcf7c.

cyberphysic4l committed:
fix iota-framework

cyberphysic4l committed:
update v2 framework snapshot

cyberphysic4l committed:
rust fmt

cyberphysic4l committed:
handle SystemEpochInfoEvents V1 and V2

miker83z commented on crates/iota-indexer/src/types.rs:
```diff
@@ -146,9 +167,36 @@ impl IndexedEpochInfo {
+        event: &SystemEpochInfoEvent,
         network_total_tx_num_at_last_epoch_end: u64,
     ) -> IndexedEpochInfo {
+        let storage_charge: Option<u64>;
+        let storage_rebate: Option<u64>;
+        let total_gas_fees: Option<u64>;
+        let total_stake_rewards_distributed: Option<u64>;
+        let burnt_tokens_amount: Option<u64>;
+        let minted_tokens_amount: Option<u64>;
+        let tips_amount: Option<u64>;
+        match event {
+            SystemEpochInfoEvent::V1(event) => {
+                storage_charge = Some(event.storage_charge);
+                storage_rebate = Some(event.storage_rebate);
+                total_gas_fees = Some(event.total_gas_fees);
+                total_stake_rewards_distributed = Some(event.total_stake_rewards_distributed);
+                burnt_tokens_amount = Some(event.burnt_tokens_amount);
+                minted_tokens_amount = Some(event.minted_tokens_amount);
+                tips_amount = Some(0);
+            }
+            SystemEpochInfoEvent::V2(event) => {
+                storage_charge = Some(event.storage_charge);
+                storage_rebate = Some(event.storage_rebate);
+                total_gas_fees = Some(event.total_gas_fees);
+                total_stake_rewards_distributed = Some(event.total_stake_rewards_distributed);
+                burnt_tokens_amount = Some(event.burnt_tokens_amount);
+                minted_tokens_amount = Some(event.minted_tokens_amount);
+                tips_amount = Some(event.tips_amount);
+            }
+        }
```
```suggestion
        let (storage_charge, storage_rebate, total_gas_fees, total_stake_rewards_distributed, burnt_tokens_amount, minted_tokens_amount, tips_amount) = match event {
            SystemEpochInfoEvent::V1(event) => (
                Some(event.storage_charge),
                Some(event.storage_rebate),
                Some(event.total_gas_fees),
                Some(event.total_stake_rewards_distributed),
                Some(event.burnt_tokens_amount),
                Some(event.minted_tokens_amount),
                Some(0),
            ),
            SystemEpochInfoEvent::V2(event) => (
                Some(event.storage_charge),
                Some(event.storage_rebate),
                Some(event.total_gas_fees),
                Some(event.total_stake_rewards_distributed),
                Some(event.burnt_tokens_amount),
                Some(event.minted_tokens_amount),
                Some(event.tips_amount),
            ),
        }
```

miker83z commented on crates/iota-indexer/src/types.rs:
```diff
@@ -116,25 +120,42 @@ pub struct IndexedEpochInfo {
-        event: Option<&SystemEpochInfoEventV1>,
+        event: Option<&SystemEpochInfoEvent>,
     ) -> IndexedEpochInfo {
+        let total_stake: u64;
+        let storage_fund_balance: u64;
+        // NOTE: total_stake and storage_fund_balance are about new epoch,
+        // although the event is generated at the end of the previous epoch,
+        // the event is optional b/c no such event for the first epoch.
+        match event {
+            Some(SystemEpochInfoEvent::V1(e)) => {
+                total_stake = e.total_stake;
+                storage_fund_balance = e.storage_fund_balance;
+            }
+            Some(SystemEpochInfoEvent::V2(e)) => {
+                total_stake = e.total_stake;
+                storage_fund_balance = e.storage_fund_balance;
+            }
+            None => {
+                total_stake = 0;
+                storage_fund_balance = 0;
+            }
+        }
```
```suggestion
        // NOTE: total_stake and storage_fund_balance are about new epoch,
        // although the event is generated at the end of the previous epoch,
        // the event is optional b/c no such event for the first epoch.
        let (total_stake, storage_fund_balance) = match event {
            Some(SystemEpochInfoEvent::V1(e)) | Some(SystemEpochInfoEvent::V2(e)) => (e.total_stake, e.storage_fund_balance),
            None => (0, 0),
        }
```

DaughterOfMars commented on crates/iota-indexer/src/types.rs:
```diff
@@ -157,10 +205,10 @@ impl IndexedEpochInfo {
             ),
             last_checkpoint_id: Some(*last_checkpoint_summary.sequence_number()),
             epoch_end_timestamp: Some(last_checkpoint_summary.timestamp_ms),
-            storage_charge: Some(event.storage_charge),
```
Couldn't we just access these fields, and skip the match?

```rust
storage_charge: Some(event.storage_charge()),
```

cyberphysic4l replied:
you mean by adding getters to the `SystemEpochInfoEvent` enum? Do you think this would be more idiomatic/efficient? It certainly seems cleaner.

DaughterOfMars replied:
I do think it could be significantly cleaner. A lot of times patterns like this will use `enum_dispatch` and a unifying trait. Alternatively, you could have functions like this:

```rust
impl SystemEpochInfoEvent {
  pub fn storage_charge(&self) -> Option<u64> {
    match self {
      ...
    }
  }
```

Not sure if it's worth it, but it would definitely clean this particular code up

miker83z replied:
I also think it is worth it! @cyberphysic4l check how the [IotaSystemState works](https://github.com/iotaledger/iota/blob/a822eb2d1f1dc1ad7ccc4517e36413e7cc9933e7/crates/iota-types/src/iota_system_state/mod.rs#L175), it's pretty much the same scenario.

cyberphysic4l commented on crates/iota-protocol-config/src/snapshots/iota_protocol_config__test__Mainnet_version_2.snap:
```diff
@@ -0,0 +1,271 @@
+vector_swap_base_cost: 52
+debug_print_base_cost: 52
+debug_print_stack_trace_base_cost: 52
+execution_version: 2
```
@miker83z should this be v2 even without the new cut of the adapter?

miker83z replied:
Nope! We keep version 1, check the comment below

miker83z commented on crates/iota-protocol-config/src/lib.rs:
```diff
@@ -1631,14 +1642,14 @@ impl ProtocolConfig {
             match cur {
                 1 => unreachable!(),
+                2 => {
+                    cfg.execution_version = Some(2);
```
This can be removed now!

lzpap commented:
Hey, I would like to point out that `Split Cluster Check / validate-testnet (pull_request)` is failing, which means that if we deployed a node sw with this PR on the testnet/devnet that runs protocol config `1`, we would fail miserably.

Looking at the code I miss one aspect: ensuring that the node can handle past protocol versions. We can't simply add new fields to structs, or new parameters to functions if older protocol versions don't have them. One example is the creation of `advance_epoch` call in the authority. Let's imagine what would happen if you do a rolling validator update on a network where only your node has the new sw yet:
 - The system is on protocol version 1, so all other validators would construct the `advance_epoch` without the additional `computation_charge_burned` parameter.
 - This would result in a different `EndOfEpochTx` for your node, a different checkpoint hence, so you basically fork.

My point is that we need to add the changes in a way that they are backwards compatible!

+1: I'm no expert in the indexer/analytical indexer, but I'm also convinced that adding new fields to existing tables must happen via `migrations`.

miker83z commented on crates/iota-types/src/gas_model/gas_v1.rs:
```diff
@@ -398,15 +397,17 @@ mod checked {
-            // compute storage rebate, both rebate and non refundable fee
+            // compute computation cost burned and storage rebate, both rebate and non
+            // refundable fee
+            let computation_cost_burned =
```
I think here it is needed the flag check, otherwise we loose the backward compatibility with the previously executed transaction effects.

miker83z replied:
@cyberphysic4l please check this

cyberphysic4l replied:
Yes indeed, added a flag now.

cyberphysic4l committed:
Merge branch 'develop' into protocol/pcr2

cyberphysic4l committed:
adapt for systemepochinfoeventv2

cyberphysic4l commented on iota-execution/latest/iota-adapter/src/execution_engine.rs:
```diff
@@ -783,30 +784,39 @@ mod checked {
-        .map(|a| builder.input(a))
-        .collect::<Result<_, _>>();
+        ];
+        if protocol_version > 1 {
+            call_arg_vec.push(CallArg::Pure(
+                bcs::to_bytes(&params.computation_charge_burned).unwrap(),
+            ));
+        }
+        let call_arg_arguments = call_arg_vec
```
Is this ok as a way to decide whether to include the computation charges burned as an argument? Alternative is to use the "protocol_defined_base_fee" flag from the protocol config. We discussed this but I do not recall the conclusion, if there was one.

cyberphysic4l committed:
clean up after merging develop

cyberphysic4l committed:
fix reward tests

cyberphysic4l committed:
Merge branch 'develop' into protocol/pcr2

DaughterOfMars commented on Cargo.toml:
```diff
@@ -414,7 +414,7 @@ iota-rosetta = { path = "crates/iota-rosetta" }
 iota-sdk = { path = "crates/iota-sdk" }
 # core-types with json format for REST API
-iota-sdk2 = { package = "iota-rust-sdk", git = "https://github.com/iotaledger/iota-rust-sdk.git", rev = "d605da95029e74376f0f39a95526bb1a5c0ebd7a", features = ["hash", "serde", "schemars"] }
+iota-sdk2 = { package = "iota-rust-sdk", git = "ssh://git@github.com/iotaledger/iota-rust-sdk.git", rev = "d97fd865b1826fe4af79901cce9b179614a0faa7", features = ["hash", "serde", "schemars"] }
```
Why the change to SSH?

cyberphysic4l replied:
not intentional change, not sure why. Reverting that now, thanks.

cyberphysic4l committed:
formatting and max version 4

vekkiokonio committed:
Minor renaming and formatting

cyberphysic4l committed:
graphql snapshot and rename validator_target_reward

cyberphysic4l commented on crates/iota-protocol-config/src/lib.rs:
```diff
@@ -600,8 +600,8 @@ pub struct ProtocolConfig {
-    /// The number of tokens minted for validator re
-    validator_target_reward: Option<u64>,
+    /// The number of tokens minted for validator per epoch.
+    validator_subsidy: Option<u64>,
```
We can't change names of existing protocol config fields. We can only add new fields and change values. I will revert this.

cyberphysic4l committed:
revert change of protocol config naming

cyberphysic4l committed:
target_reward -> subsidy except in protocol config

cyberphysic4l committed:
remove clippy lint expectation

cyberphysic4l committed:
revert change to ssh for rust sdk

cyberphysic4l committed:
execution version 1 and protocol version 4

cyberphysic4l committed:
iota framework snapshot

cyberphysic4l committed:
Merge branch 'core-protocol/procotol-version-3' into protocol/pcr2

cyberphysic4l committed:
revert v1 execution changes

cyberphysic4l committed:
clean framework bytecode snapshot

cyberphysic4l committed:
protocol config snapshots

cyberphysic4l committed:
fix some tests

cyberphysic4l committed:
Merge branch 'develop' into protocol/pcr2

alexsporn committed:
feat: bump version to 0.10.0-alpha

alexsporn committed:
feat: bump version to 0.10.0-alpha

feat: added protocol version 4
feat: updated all e2e tests to use protocol version 4

alexsporn committed:
feat: bump version to 0.10.0-alpha

feat: added protocol version 4
feat: updated all e2e tests to use protocol version 4

alexsporn committed:
feat: added protocol version 4

alexsporn committed:
feat: updated all e2e tests to use protocol version 4

cyberphysic4l committed:
Merge remote-tracking branch 'origin/core-protocol/procotol-version-4' into protocol/pcr2

cyberphysic4l committed:
fix move_package_upgrade_tests

cyberphysic4l committed:
fix pay_iota_tests

cyberphysic4l committed:
fix bridge tests

cyberphysic4l committed:
Merge branch 'core-protocol/procotol-version-4' into protocol/pcr2

cyberphysic4l committed:
update iota-adapter-transactional-tests baseline

cyberphysic4l committed:
update openapi.json

cyberphysic4l committed:
update iota-swarm-config snapshot

vekkiokonio committed:
Fix checkpoint table

vekkiokonio committed:
Update test baselines

cyberphysic4l committed:
fix production upgrade test

cyberphysic4l committed:
update basic_read_cmd_snapshot_tests snapshot

cyberphysic4l committed:
fix safe_mode_reconfig_test

cyberphysic4l committed:
remove unneccesary SystemEpochInfoEvent enum from framework

cyberphysic4l committed:
regenerate light client checkpoints

cyberphysic4l committed:
clean snapshot manifest

cyberphysic4l committed:
update iota-swarm-config snapshot

cyberphysic4l committed:
update graphql e2e tests

cyberphysic4l committed:
fix protocol version tests

cyberphysic4l committed:
create ChangeEpochV2

cyberphysic4l committed:
update graphql rpc

cyberphysic4l committed:
update iota-rust-sdk

cyberphysic4l committed:
rust fmt fix

cyberphysic4l committed:
Revert "regenerate light client checkpoints"

This reverts commit c83cd546aa5f507485e146803ca6be0bb64c9313.

cyberphysic4l committed:
update openrpc

cyberphysic4l committed:
fix graphql e2e tests

cyberphysic4l committed:
update iota-core format

cyberphysic4l committed:
update iota-rest-api openapi.json

alexsporn commented:
Please rebase on top of the latest rebased `core-protocol/procotol-version-4` branch

alexsporn committed:
Merge branch 'core-protocol/procotol-version-4' of github.com:iotaledger/iota into protocol/pcr2

miker83z commented on crates/iota-types/src/iota_system_state/iota_system_state_summary.rs:
```diff
@@ -75,7 +75,11 @@ pub struct IotaSystemStateSummary {
     #[schemars(with = "BigInt<u64>")]
     #[serde_as(as = "Readable<BigInt<u64>, _>")]
-    pub safe_mode_computation_rewards: u64,
+    pub safe_mode_computation_charges: u64,
```
Pls @cyberphysic4l, make sure that @iotaledger/infrastructure @iotaledger/tooling @iotaledger/dev-tools are aware of this change when this PR is finalized.

miker83z commented on iota-execution/latest/iota-adapter/src/execution_engine.rs:
```diff
@@ -845,7 +868,7 @@ mod checked {
     fn advance_epoch(
         builder: ProgrammableTransactionBuilder,
-        change_epoch: ChangeEpoch,
+        change_epoch: ChangeEpochV2,
```
- Rename `advance_epoch` to `advance_epoch_v1` and keep the `ChangeEpoch` type as parameter; this function will call `construct_advance_epoch_pt`;
- Create `advance_epoch_v2` that uses `ChangeEpochV2` as parameter; this function will call `construct_advance_epoch_pt_v2`;
- Create a function  `advance_epoch_impl` that is called by `advance_epoch_v1` and `advance_epoch_v2` for the code part that is in common (execution and result handling).

miker83z commented on iota-execution/latest/iota-adapter/src/execution_engine.rs:
```diff
@@ -783,30 +798,38 @@ mod checked {
     pub fn construct_advance_epoch_pt(
         mut builder: ProgrammableTransactionBuilder,
         params: &AdvanceEpochParams,
+        protocol_version: u64,
```
- Remove this protocol_version
- Create a `construct_advance_epoch_pt_v1` that works as before (i.e., with no `computation_charge_burned`) to ba called by `advance_epoch_v1` (check below)
- Create a `construct_advance_epoch_pt_v2` that is called only by `advance_epoch_v2` and uses  `computation_charge_burned`
- Create a function `construct_advance_epoch_pt_impl` that is called by `construct_advance_epoch_pt_v1` and `construct_advance_epoch_pt_v2` for the code part that is in common (move calls and builder).

miker83z commented on iota-execution/latest/iota-adapter/src/execution_engine.rs:
```diff
@@ -663,6 +664,20 @@ mod checked {
                     match tx {
                         EndOfEpochTransactionKind::ChangeEpoch(change_epoch) => {
+                            assert_eq!(i, len - 1);
+                            advance_epoch(
```
Invoke `advance_epoch_v1` (see below)

miker83z commented on iota-execution/latest/iota-adapter/src/execution_engine.rs:
```diff
@@ -663,6 +664,20 @@ mod checked {
+                            )?;
+                            return Ok(Mode::empty_results());
+                        }
+                        EndOfEpochTransactionKind::ChangeEpochV2(change_epoch) => {
```
Invoke `advance_epoch_v2` (see below)

miker83z commented on iota-execution/latest/iota-adapter/src/execution_engine.rs:
```diff
@@ -913,7 +938,7 @@ mod checked {
 
     fn process_system_packages(
-        change_epoch: ChangeEpoch,
+        change_epoch: ChangeEpochV2,
```
Same as above: `process_system_packages_v1`, `process_system_packages_v2` and `process_system_packages_impl`

miker83z replied:
To the impl you could pass the result of `change_epoch.system_packages.into_iter()` directly instead of the `ChangeEpoch` type

cyberphysic4l replied:
When I pass the system_packages directly, I don't need the different versions either, just the impl, so even better.

miker83z reviewed (changes requested):
I think it is inevitable to create a ChangeEpochV2, but the executor could be refactored:

cyberphysic4l committed:
use flag for calculating gascostsummary

cyberphysic4l committed:
refactor executiong engine

alexsporn commented on crates/iota-genesis-builder/src/lib.rs:
```diff
@@ -521,9 +521,9 @@ impl Builder {
-        #[expect(clippy::infallible_destructuring_match)]
         let system_state = match unsigned_genesis.iota_system_object() {
             IotaSystemState::V1(inner) => inner,
+            IotaSystemState::V2(_) => unreachable!(),
```
@miker83z is this correct here? If a genesis is created starting with protocol version 4 should't the IotaSystemState be a V2 object from the start?

miker83z replied:
Yes, correct. 

The [genesis::create](https://github.com/iotaledger/iota/blob/a02ccdab1fbdcbfed5a1a0f9685b3374571e3d5c/crates/iota-framework/packages/iota-system/sources/genesis.move#L81) rules here. It's a method that shouldn't change normally and if you check its behavior, it calls [iota_system::create](https://github.com/iotaledger/iota/blob/a02ccdab1fbdcbfed5a1a0f9685b3374571e3d5c/crates/iota-framework/packages/iota-system/sources/iota_system.move#L73) which always create a `IotaSystemV1`.  This makes the new framework upgrades backward compatible with the original genesis building. What happens then is that, after the genesis is created, the first time the `IotaSystem` is invoked in a transaction, it inevitably calls `load_system_state_mut` and then [load_inner_maybe_upgrade](https://github.com/iotaledger/iota/blob/a02ccdab1fbdcbfed5a1a0f9685b3374571e3d5c/crates/iota-framework/packages/iota-system/sources/iota_system.move#L561), which immediately converts IotaSystemV1 to V2.

alexsporn commented on crates/iota-indexer/migrations/pg/2023-08-19-044044_checkpoints/up.sql:
```diff
@@ -13,6 +13,7 @@ CREATE TABLE checkpoints
     timestamp_ms                        BIGINT       NOT NULL,
     total_gas_cost                      BIGINT       NOT NULL,
     computation_cost                    BIGINT       NOT NULL,
+    computation_cost_burned             BIGINT       NOT NULL,
```
This probably needs to go into a separate migration.
@kodemartin please check

muXxer replied:
I agree.

I wonder if all these migration files are actually needed, because we started with a much cleaner state anyway from the beginning.

For this PR we need a new migration file I guess.

alexsporn commented on crates/iota-types/src/transaction.rs:
```diff
@@ -307,21 +335,36 @@ impl EndOfEpochTransactionKind {
-            epoch_start_timestamp_ms,
-            system_packages,
-        })
+        if protocol_version.as_u64() > 3 {
```
I don't like this solution. Hardcoding protocol versions like this is very intransparent.
Better have 2 initializers and switch on the call-site depending on a protocol-config parameter.

cyberphysic4l committed:
add some move tests for tips

cyberphysic4l committed:
use feature flag instead of protocol version for selecting ChangeEpoch kind

cyberphysic4l committed:
Merge branch 'develop' into protocol/pcr2

alexsporn committed:
Merge branch 'core-protocol/procotol-version-4' into protocol/pcr2

cyberphysic4l committed:
fmt

cyberphysic4l committed:
add computation_cost_burned to indexer checkpoint schema

muXxer commented on schema.graphql:
```diff
@@ -0,0 +1,4651 @@
+type ActiveJwk {
```
this file shouldn't be here in the root path at all 😅

muXxer commented on external-crates/move/Cargo.lock:
```diff
@@ -1454,6 +1454,15 @@ dependencies = [
 ]
 
+[[package]]
+name = "move-abstract-interpreter-v1"
```
how come that this file changed at all?

cyberphysic4l replied:
I think this is left over from a previous iteration where we planned to do a new execution cut.

muXxer commented on crates/iota-core/src/authority.rs:
```diff
@@ -4654,15 +4670,28 @@ impl AuthorityState {
             .map(|event| {
-                bcs::from_bytes::<SystemEpochInfoEventV1>(&event.contents)
-                    .expect("event deserialization should succeed as type was pre-validated")
+                if event.is_system_epoch_info_event_v1() {
```
Not a biggie, but personally I would check for the happy case first, which is always the lastest version in our case.

cyberphysic4l replied:
fair enough, changed.

muXxer commented on crates/iota-core/src/unit_tests/pay_iota_tests.rs:
```diff
@@ -70,7 +70,7 @@ async fn test_pay_iota_failure_insufficient_gas_balance_one_input_coin() {
 async fn test_pay_iota_failure_insufficient_total_balance_one_input_coin() {
     let (sender, sender_key): (_, AccountKeyPair) = get_key_pair();
-    let coin1 = Object::with_id_owner_gas_for_testing(ObjectID::random(), sender, 500100);
+    let coin1 = Object::with_id_owner_gas_for_testing(ObjectID::random(), sender, 1000100);
```
was it necessary to actually change the values in this file?
if not, it increases the "noise" of this PR review and if we ever want compare to upstream :D

cyberphysic4l replied:
Yes it was necessary due to the change from reference gas price (500) with to base gas price from the protocol config (1000)

muXxer commented on crates/iota-protocol-config/src/lib.rs:
```diff
@@ -188,6 +189,10 @@ struct FeatureFlags {
 
+    // Enable a protocol-defined base gas price for all transactions.
+    #[serde(skip_serializing_if = "is_false")]
+    protocol_defined_base_fee: bool,
```
We should add this new flag at the end of the list.

cyberphysic4l replied:
Agreed, done.

muXxer commented on crates/iota-protocol-config/src/lib.rs:
```diff
@@ -1052,6 +1059,10 @@ impl ProtocolConfig {
         self.feature_flags.passkey_auth
     }
 
+    pub fn protocol_defined_base_fee(&self) -> bool {
```
Please also move the getter to the end of the functions, or at least after `relocate_event_module`

cyberphysic4l replied:
Done.

muXxer commented on crates/iota-types/src/event.rs:
```diff
@@ -177,3 +189,21 @@ pub struct SystemEpochInfoEventV1 {
+
+// Event emitted in move code `fun advance_epoch` in protocol versions 4 and
+// later
+#[derive(Deserialize)]
```
I would add a comment about what fields were added or removed in this new version, and why it was necessary to create a new type.
It's easier to understand if you read the code and wonder why this type was introduced.

muXxer commented on crates/iota-types/src/iota_system_state/iota_system_state_inner_v2.rs:
```diff
@@ -0,0 +1,280 @@
+        // If you are making any changes to IotaSystemStateV1 or any of its dependent
+        // types before mainnet, please also update IotaSystemStateSummary and
+        // its corresponding TS type. Post-mainnet, we will need to introduce a
+        // new version.
```
What is that weird comment :D
Can we remove that? Maybe also from v1?

muXxer commented on crates/iota-types/src/iota_system_state/iota_system_state_inner_v2.rs:
```diff
@@ -0,0 +1,280 @@
+
+/// Rust version of the Move iota_system::iota_system::IotaSystemStateV1 type
+#[derive(Debug, Serialize, Deserialize, Clone, Eq, PartialEq)]
+pub struct IotaSystemStateV2 {
```
I would add a comment about what fields were added or removed in this new version, and why it was necessary to create a new type.
It's easier to understand if you read the code and wonder why this type was introduced.

gokhan-simsek-iota replied:
I agree, this is how I did that in my draft PR (but still thinking about the best location of that comment): 
"    /// V2: primary_address, next_epoch_primary_address are enforced to be TCP "
https://github.com/iotaledger/iota/pull/4753/commits/0264d58e95352ba1351b185d093960f95714a6c2#diff-2fa17cd033ca5ad615b4db77ecee29f946d54dde5fd5e40a471f0966fd158580R54

muXxer commented on crates/iota-types/src/transaction.rs:
```diff
@@ -58,7 +58,7 @@ use crate::{
 pub const TEST_ONLY_GAS_UNIT_FOR_TRANSFER: u64 = 10_000;
 pub const TEST_ONLY_GAS_UNIT_FOR_OBJECT_BASICS: u64 = 50_000;
-pub const TEST_ONLY_GAS_UNIT_FOR_PUBLISH: u64 = 70_000;
+pub const TEST_ONLY_GAS_UNIT_FOR_PUBLISH: u64 = 50_000;
```
why the change?

cyberphysic4l replied:
This is to reduce the computation units used in some tests accomodate the increase from reference gas price with the base gas price of 1000 from protocol config rather than the standard rgp of 700 from the gas price survey.

muXxer commented on crates/iota-types/src/transaction.rs:
```diff
@@ -190,6 +190,33 @@ pub struct ChangeEpoch {
 }
 
+#[derive(Debug, PartialEq, Eq, Hash, Clone, Serialize, Deserialize)]
+pub struct ChangeEpochV2 {
```
I would add a comment about what fields were added or removed in this new version, and why it was necessary to create a new type.
It's easier to understand if you read the code and wonder why this type was introduced.

muXxer commented on crates/iota-types/src/transaction.rs:
```diff
@@ -295,6 +322,7 @@ pub enum TransactionKind {
 #[derive(Debug, PartialEq, Eq, Hash, Clone, Serialize, Deserialize, IntoStaticStr)]
 pub enum EndOfEpochTransactionKind {
     ChangeEpoch(ChangeEpoch),
+    ChangeEpochV2(ChangeEpochV2),
```
Stupid question, maybe @thibault-martinez knows.
Does it matter for serialization in which position the new enum value is put? Do we break anything with this?
As far as I see they were always put at the end on upstream, but this one looks cleaner.

muXxer commented on crates/iota-e2e-tests/tests/framework_upgrades/mock_iota_systems/base/sources/iota_system.move:
```diff
@@ -49,16 +49,17 @@ module iota_system::iota_system {
         non_refundable_storage_fee: u64,
         reward_slashing_rate: u64,
         epoch_start_timestamp_ms: u64,
+        computation_charge_burned: u64,
```
If you change the function signature anyway, why not directly after `computation_charge`.

cyberphysic4l replied:
This was because in a previous iteration, we added the final additional argument to a vector of the other arguments and it was cleanest to have the new argument last. I can revert this now as you suggest because the approach has changed.

muXxer commented on crates/iota-e2e-tests/tests/framework_upgrades/mock_iota_systems/deep_upgrade/sources/iota_system.move:
```diff
@@ -49,16 +49,17 @@ module iota_system::iota_system {
         non_refundable_storage_fee: u64,
         reward_slashing_rate: u64,
         epoch_start_timestamp_ms: u64,
+        computation_charge_burned: u64,
```
If you change the function signature anyway, why not directly after `computation_charge`.

muXxer commented on crates/iota-e2e-tests/tests/framework_upgrades/mock_iota_systems/safe_mode/sources/iota_system.move:
```diff
@@ -49,25 +49,27 @@ module iota_system::iota_system {
         non_refundable_storage_fee: u64,
         reward_slashing_rate: u64,
         _epoch_start_timestamp_ms: u64,
+        computation_charge_burned: u64,
```
If you change the function signature anyway, why not directly after `computation_charge`.

muXxer commented on crates/iota-e2e-tests/tests/framework_upgrades/mock_iota_systems/safe_mode/sources/iota_system_state_inner.move:
```diff
@@ -92,9 +93,10 @@ module iota_system::iota_system_state_inner {
-        _validator_target_reward: u64,
+        _validator_subsidy: u64,
         mut storage_charge: Balance<IOTA>,
         mut computation_reward: Balance<IOTA>,
```
Isn't that the `computation_charge` now?

cyberphysic4l replied:
Yes, I haven't bothered to rename for the mock system states, but I can do that.

muXxer commented on crates/iota-e2e-tests/tests/framework_upgrades/mock_iota_systems/shallow_upgrade/sources/iota_system.move:
```diff
@@ -49,16 +49,17 @@ module iota_system::iota_system {
         non_refundable_storage_fee: u64,
         reward_slashing_rate: u64,
         epoch_start_timestamp_ms: u64,
+        computation_charge_burned: u64,
```
If you change the function signature anyway, why not directly after `computation_charge`.

muXxer commented on crates/iota-indexer/src/handlers/checkpoint_handler.rs:
```diff
@@ -226,20 +226,33 @@ impl CheckpointHandler {
-            .find(|ev| ev.is_system_epoch_info_event())
+            .find(|ev| ev.is_system_epoch_info_event_v1() || ev.is_system_epoch_info_event_v2())
+            .map(|ev| {
+                if ev.is_system_epoch_info_event_v1() {
```
Again the happy case thing. I would check first for the new version.

muXxer commented on crates/iota-types/src/gas_model/gas_v1.rs:
```diff
@@ -397,15 +403,20 @@ mod checked {
+            let computation_cost_burned = if self.protocol_defined_base_fee {
+                self.computation_cost * self.reference_gas_price / self.gas_price
+            } else {
+                self.computation_cost
```
```suggestion
                // if protocol_defined_base_fee is not active, the entire computation cost is burned.
                self.computation_cost
```

muXxer replied:
It would also be nice to add a comment to `self.computation_cost * self.reference_gas_price / self.gas_price`

muXxer commented on crates/iota-types/src/gas_model/gas_v1.rs:
```diff
@@ -371,24 +380,21 @@ mod checked {
-                } else {
-                    ((gas_used / gas_rounding) + 1) * gas_rounding * self.gas_price
+            let mut computation_units = self.gas_status.gas_used_pre_gas_price();
+            if let Some(gas_rounding) = self.gas_rounding_step {
```
Is it ever checked if gas_rounding can't be zero?
Otherwise this is a bit nasty 😅

cyberphysic4l replied:
Added a check because it seems it would be technically possible to set gas_rounding_step to Some(0).

muXxer commented on crates/iota-types/src/gas_model/gas_v1.rs:
```diff
@@ -371,24 +380,21 @@ mod checked {
                 }
             } else {
-                let bucket_cost = get_bucket_cost(&self.cost_table.computation_bucket, gas_used);
-                // charge extra on top of `computation_cost` to make the total computation
```
is the comment still valid and was removed by accident?

cyberphysic4l replied:
It is still valid, I added it back in.

muXxer commented on crates/iota-framework/packages/iota-system/sources/iota_system.move:
```diff
@@ -511,22 +511,23 @@ module iota_system::iota_system {
         non_refundable_storage_fee: u64,
         reward_slashing_rate: u64, // how much rewards are slashed to punish a validator, in bps.
         epoch_start_timestamp_ms: u64, // Timestamp of the epoch start
+        computation_charge_burned: u64,
```
If you change the function signature anyway, why not directly after `computation_charge`.

muXxer commented on crates/iota-framework/packages/iota-system/sources/iota_system.move:
```diff
@@ -727,28 +738,30 @@ module iota_system::iota_system {
             non_refundable_storage_fee,
             reward_slashing_rate,
             epoch_start_timestamp_ms,
+            computation_charge_burned,
```
If you change the function signature anyway, why not directly after `computation_charge`.

muXxer commented on crates/iota-framework/packages/iota-system/sources/iota_system_state_inner.move:
```diff
@@ -199,6 +270,50 @@ module iota_system::iota_system_state_inner {
         }
     }
 
+    public(package) fun v1_to_v2(self: IotaSystemStateV1): IotaSystemStateV2 {
```
Not sure if we should call it this way.

I see that on upstream it is the same name, but sometimes they also use `the_struct_to_update_v1_to_v2`, like in `validator_set_v1_to_v2`. The function will be publicly visible, so maybe we should be more explicit? What does @iotaledger/vm-language say?

miker83z replied:
I would keep the same pattern as upstream! Moreover, the function is publicly visible only to the package, so no external devs will interact with this.

muXxer replied:
alright!

muXxer commented on crates/iota-framework/packages/iota-system/sources/iota_system_state_inner.move:
```diff
@@ -108,8 +108,60 @@ module iota_system::iota_system_state_inner {
-    /// Event containing system-level epoch information, emitted during
-    /// the epoch advancement transaction.
+    /// The top-level object containing all information of the Iota system.
+    public struct IotaSystemStateV2 has store {
```
I would add a comment about what fields were added or removed in this new version, and why it was necessary to create a new type.
It's easier to understand if you read the code and wonder why this type was introduced.

muXxer commented on crates/iota-framework/packages/iota-system/sources/iota_system_state_inner.move:
```diff
@@ -124,6 +176,25 @@ module iota_system::iota_system_state_inner {
+    #[allow(unused_field)]
+    /// The second version of the event containing system-level epoch information,
+    /// emitted during the epoch advancement transaction.
+    public struct SystemEpochInfoEventV2 has copy, drop {
```
I would add a comment about what fields were added or removed in this new version, and why it was necessary to create a new type.
It's easier to understand if you read the code and wonder why this type was introduced.

muXxer commented on crates/iota-graphql-rpc/src/types/transaction_block_kind/end_of_epoch.rs:
```diff
@@ -43,6 +44,7 @@ pub(crate) struct EndOfEpochTransaction {
 #[derive(Union, Clone, PartialEq, Eq)]
 pub(crate) enum EndOfEpochTransactionKind {
     ChangeEpoch(ChangeEpochTransaction),
+    ChangeEpochV2(ChangeEpochTransactionV2),
```
Same question about the position in the enum here.

muXxer commented on crates/iota-graphql-rpc/src/types/transaction_block_kind/end_of_epoch.rs:
```diff
@@ -56,6 +58,13 @@ pub(crate) struct ChangeEpochTransaction {
 }
 
+#[derive(Clone, PartialEq, Eq)]
+pub(crate) struct ChangeEpochTransactionV2 {
```
Please add the comment that here the `computation_charge_burned` was added.

muXxer commented on crates/iota-json-rpc-types/src/iota_transaction.rs:
```diff
@@ -1613,6 +1657,7 @@ pub struct IotaEndOfEpochTransaction {
 #[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
 pub enum IotaEndOfEpochTransactionKind {
     ChangeEpoch(IotaChangeEpoch),
+    ChangeEpochV2(IotaChangeEpochV2),
```
Same question here about the enum position.

cyberphysic4l committed:
Mux review

cyberphysic4l committed:
update framework snapshot

cyberphysic4l committed:
update test baselines
